### PR TITLE
Remove use of CUDA driver API calls from libcudf

### LINKED ISSUE #3661: [FEA] Import cudf on non-CUDA enabled machine
I would like to be able to `import cudf` and refer to cudf methods and classes on a machine that does not have a GPU.  

This is particularly useful when using cudf with Dask.  While my local machine may not have an attached GPU, my Dask workers may.   When using Dask with cudf I need to be able to refer to cudf functions so that I can place them into a task graph, but don't actually need to run them locally.  

This fails today with an import error saying that we can't import `libcuda.so.1`.  Is it feasible to make importing cudf robust to these errors?

### PR DESCRIPTION
**Pull request by shwina**
## Description

This PR adds the final bits that remove `libcuda.so` as a shared library dependency of `libcudf`. Currently, that's just one place that uses the driver API `cuContextGetCurrent()` to get the current CUDA context and maintain a per-context singleton. It was determined in the discussions within this PR that that could be replaced with a per-device singleton and a call to the runtime API `cudaGetDevice()` instead. 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/cudf/strings/detail/utilities.cuh [hunk 1]
```diff
@@ -195,8 +195,17 @@ class per_context_cache {
...
   TableType* find_or_initialize(const Initializer& init)
   {
+    void* cuda_lib                                 = nullptr;
+    decltype(cuCtxGetCurrent)* context_get_current = nullptr;
+
+    cuda_lib = dlopen("libcuda.so", RTLD_LAZY | RTLD_LOCAL | RTLD_NODELETE);
+    CUDF_EXPECTS(cuda_lib != nullptr, "Failed to load cuda library");
+    context_get_current =
+      reinterpret_cast<decltype(context_get_current)>(dlsym(cuda_lib, "cuCtxGetCurrent"));
+    CUDF_EXPECTS(context_get_current != nullptr, "could not find cuFile cuFileDriverOpen symbol");
```
Comment (shwina) [inline cpp/include/cudf/strings/detail/utilities.cuh:205]:
```suggestion
    CUDF_EXPECTS(context_get_current != nullptr, "could not find cuCtxGetCurrent symbol");
```

File: cpp/include/cudf/strings/detail/utilities.cuh [hunk 2]
```diff
@@ -25,13 +25,13 @@
 #include <rmm/cuda_stream_view.hpp>
 #include <rmm/exec_policy.hpp>
 
+#include <dlfcn.h>
+#include <mutex>
```
Comment (davidwendt) [inline cpp/include/cudf/strings/detail/utilities.cuh:29]: Can  you move these includes down next to `#include <unordered_map>` ?

**Review by davidwendt — CHANGES_REQUESTED** [review]

File: cpp/include/cudf/strings/detail/utilities.cuh (hunk 2 above)
Comment (shwina) [reply cpp/include/cudf/strings/detail/utilities.cuh:29]: `clang-format` seems to want to pull them back up here?

**Review by robertmaynard — APPROVED** [review]

File: cpp/include/cudf/strings/detail/utilities.cuh (hunk 2 above)
Comment (davidwendt) [reply cpp/include/cudf/strings/detail/utilities.cuh:29]: The empty line is important.
```
#include <thrust/distance.h>
#include <thrust/for_each.h>
#include <thrust/iterator/counting_iterator.h>
#include <thrust/iterator/transform_iterator.h>
#include <thrust/scan.h>

#include <dlfcn.h>
#include <mutex>
#include <unordered_map>
```

Comment (shwina) [reply cpp/include/cudf/strings/detail/utilities.cuh:29]: Thanks! Fix'd;

File: cpp/include/cudf/strings/detail/utilities.cuh [hunk 3]
```diff
@@ -195,8 +195,17 @@ class per_context_cache {
...
   TableType* find_or_initialize(const Initializer& init)
   {
+    void* cuda_lib                                 = nullptr;
+    decltype(cuCtxGetCurrent)* context_get_current = nullptr;
+
+    cuda_lib = dlopen("libcuda.so", RTLD_LAZY | RTLD_LOCAL | RTLD_NODELETE);
+    CUDF_EXPECTS(cuda_lib != nullptr, "Failed to load cuda library");
+    context_get_current =
+      reinterpret_cast<decltype(context_get_current)>(dlsym(cuda_lib, "cuCtxGetCurrent"));
+    CUDF_EXPECTS(context_get_current != nullptr, "could not find cuCtxGetCurrent symbol");
```
Comment (jrhemstad) [inline cpp/include/cudf/strings/detail/utilities.cuh:205]: I strongly suspect we can eliminate the need for using the driver API at all here by just using `cudaGetDevice` instead.

The only reason for using `cuCtxGetCurrent` would be if someone is manually constructing and managing a custom `CUcontext` and I strongly suspect no one using libcudf is doing that. The exception could be Legion/Realm given @magnatelee is the one who originally wrote this here: https://github.com/rapidsai/cudf/issues/5335

**Review by davidwendt — APPROVED** [review]

File: cpp/include/cudf/strings/detail/utilities.cuh (hunk 3 above)
Comment (magnatelee) [reply cpp/include/cudf/strings/detail/utilities.cuh:206]: @jrhemstad I think that code should work equally well with `cudaGetDevice`.

Comment (davidwendt) [reply cpp/include/cudf/strings/detail/utilities.cuh:206]: Thanks Wonchan Lee. I can work on changing this code to use `cudaGetDevice` instead in a separate PR to not hold up this fix. Would that be ok @jrhemstad ?

Comment (shwina) [reply cpp/include/cudf/strings/detail/utilities.cuh:206]: If you just want to take over  this PR @davidwendt , that  would be great as well!

**Review by jrhemstad — APPROVED** [review]
I'm fine with merging this PR for now and switching to `cudaGetDevice` in a later PR.

**Review by magnatelee — APPROVED** [review]

File: cpp/include/cudf/strings/detail/utilities.cuh [hunk 4]
```diff
@@ -31,6 +31,7 @@
 #include <thrust/iterator/transform_iterator.h>
 #include <thrust/scan.h>
 
+#include <dlfcn.h>
```
Comment (shwina) [inline cpp/include/cudf/strings/detail/utilities.cuh:34]:
```suggestion
```

**Comment by shwina** [comment]
@gpucibot merge

**Pull request merged**